record_linux/lib/record_linux.dart

```diff
@@ -65,76 +66,82 @@ class RecordLinux extends RecordPlatform {
-
-    final file = File(path);
-    if (file.existsSync()) await file.delete();
+@override
+Future<void> start(
+  String recorderId,
+  RecordConfig config, {
+  required String path,
+}) async {
+  await stop(recorderId);
 
-    final supported = await isEncoderSupported(recorderId, config.encoder);
-    if (!supported) {
-      throw Exception('${config.encoder} is not supported.');
-    }
+  final file = File(path);
+  if (file.existsSync()) await file.delete();
 
-    String numChannels;
-    if (config.numChannels == 6) {
-      numChannels = '5.1';
-    } else if (config.numChannels == 8) {
-      numChannels = '7.1';
-    } else if (config.numChannels == 1 || config.numChannels == 2) {
-      numChannels = config.numChannels.toString();
-    } else {
-      throw Exception('${config.numChannels} config is not supported.');
-    }
+  final supported = await isEncoderSupported(recorderId, config.encoder);
+  if (!supported) {
+    throw Exception('${config.encoder} is not supported.');
+  }
 
-    await _callFMedia(
-      [
-        '--notui',
-        '--background',
-        '--record',
-        '--out=$path',
-        '--rate=${config.sampleRate}',
-        '--channels=$numChannels',
-        '--globcmd=listen',
-        '--gain=6.0',
-        if (config.device != null) '--dev-capture=${config.device!.id}',
-        ..._getEncoderSettings(config.encoder, config.bitRate),
-      ],
-      onStarted: () {
-        _path = path;
-        _updateState(RecordState.record);
-      },
-      consumeOutput: false,
-      recorderId: recorderId,
-    );
+  String numChannels;
+  if (config.numChannels == 1 || config.numChannels == 2) {
+    numChannels = config.numChannels.toString();
+  } else {
+    throw Exception('${config.numChannels} config is not supported.');
   }
 
+  final args = [
+    '--raw',
+    '--format=s16le',
+    '--rate=${config.sampleRate}',
+    '--channels=$numChannels',
+    '--latency-msec=100',
+    if (config.device != null) '--device=${config.device!.id}',
+    if (config.autoGain) '--property=auto_gain_control=1',
+    if (config.echoCancel) '--property=echo_cancellation=1',
+    if (config.noiseSuppress) '--property=noise_suppression=1',
+  ];
+
+  _parecordProcess = await Process.start(_parecordBin, args);
+
+  final ffmpegArgs = [
+    '-f',
+    's16le',
+    '-ar',
+    config.sampleRate.toString(),
+    '-ac',
+    numChannels,
+    '-i',
+    '-',
+    ..._getEncoderSettings(config.encoder, path, config.bitRate)
+  ];
+
+  _ffmpegProcess = await Process.start(_ffmpegBin, ffmpegArgs);
+
+  _parecordProcess!.stdout.pipe(_ffmpegProcess!.stdin);
+
+  _path = path;
+  _updateState(RecordState.record);
+}
+
   @override
   Future<String?> stop(String recorderId) async {
     final path = _path;
 
-    await _callFMedia(['--globcmd=stop'], recorderId: recorderId);
-    await _callFMedia(['--globcmd=quit'], recorderId: recorderId);
+    _parecordProcess?.kill();
+    _parecordProcess = null;
```

Comment:
How ffmpeg process is stopped?
How to ensure that ffmpeg process has finished before sending state update?

Reply:
since ffmpeg is being piped, it will stop on its own as soon as it finishes the processing of its buffer. Force terminating it earlier can be risky. I will however set it back to null which I missed!